Tests/PostgresNIOTests/New/PostgresConnectionTests.swift

```diff
@@ -638,6 +638,63 @@ class PostgresConnectionTests: XCTestCase {
         }
     }
 
+    func testQueriesFailIfConnectionIsClosed() async throws {
+        let (connection, channel) = try await self.makeTestConnectionWithAsyncTestingChannel()
+
+        try await connection.closeGracefully()
+
+        XCTAssertEqual(channel.isActive, false)
+
+        do {
+            _ = try await connection.query("SELECT version;", logger: self.logger)
+            XCTFail("Expected to fail")
+        } catch let error as ChannelError {
+            XCTAssertEqual(error, .ioOnClosedChannel)
+        }
+    }
+
+    func testListenFailsIfConnectionIsClosed() async throws {
+        let (connection, channel) = try await self.makeTestConnectionWithAsyncTestingChannel()
+
+        try await connection.closeGracefully()
+
+        XCTAssertEqual(channel.isActive, false)
+
+        do {
+            _ = try await connection.listen("test_channel")
+            XCTFail("Expected to fail")
+        } catch let error as ChannelError {
+            XCTAssertEqual(error, .ioOnClosedChannel)
+        }
+    }
+
+    func testListenFailsIfConnectionIsClosedMidway() async throws {
+        let (connection, channel) = try await self.makeTestConnectionWithAsyncTestingChannel()
+
+        try await withThrowingTaskGroup(of: Void.self) { taskGroup in
+            taskGroup.addTask {
+                do {
+                    let listenSequence = try await connection.listen("test_channel")
+
+                    for try await _ in listenSequence {
+                        XCTFail("Expected to fail")
+                    }
+                } catch let error as PSQLError {
+                    XCTAssertEqual(error.code, .listenFailed)
+                }
+            }
+
+            taskGroup.addTask {
+                try await Task.sleep(for: .seconds(3))
+
+                try await connection.close().get()
+                XCTAssertEqual(channel.isActive, false)
+            }
+
+            try await taskGroup.waitForAll()
+        }
+    }
```

Comment:
@fabianfett This is related to #458?
I noticed this same issue before they file an issue for when i was trying to put a `ServiceDiscovery` together using PostgresNIO for `DistributedActor`s, but didn't get the chance/mood to file and issue or it before they did.

Reply:
revamped this test so this discussion is no longer valid. Though https://github.com/vapor/postgres-nio/pull/487#discussion_r1649799205 is related to the crash.